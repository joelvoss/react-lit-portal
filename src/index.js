import * as React from 'react';
import { createPortal } from 'react-dom';
import {
	useIsomorphicLayoutEffect as useLayoutEffect,
	useForceUpdate,
} from '@react-lit/helper';

/**
 * Portal appends a new DOM node to the end of the `document.body` and renders
 * it's child React tree into it. Useful to break out of the DOM hierarchy to
 * prevent parent styles from clipping or disturbing content (e.g. popovers,
 * dropdowns and modals).
 */
export function Portal({ children, type = 'portal' }) {
	/** @type {React.MutableRefObject<HTMLDivElement | null>} */
	let mountNode = React.useRef(null);
	/** @type {React.MutableRefObject<HTMLElement | null>} */
	let portalNode = React.useRef(null);
	let forceUpdate = useForceUpdate();

	useLayoutEffect(() => {
		// NOTE(joel): Guard against undefined ref's in case of hot-loader
		// component replacements.
		if (!mountNode.current) return;

		// NOTE(joel): Make sure we're appending to the correct document element.
		const ownerDocument = mountNode.current.ownerDocument;
		portalNode.current = ownerDocument?.createElement(type);
		ownerDocument.body.appendChild(portalNode.current);

		forceUpdate();

		return () => {
			if (portalNode.current && portalNode.current.ownerDocument) {
				portalNode.current.ownerDocument.body.removeChild(portalNode.current);
			}
		};
	}, [type, forceUpdate]);

	return portalNode.current ? (
		createPortal(children, portalNode.current)
	) : (
		<span ref={mountNode} />
	);
}
