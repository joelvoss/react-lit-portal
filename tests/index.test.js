import * as React from 'react';
import { render } from './test-utils';

import { Portal } from '../src/index';

describe('<Portal />', () => {
	it(`should render a portal`, async () => {
		const Element = () => (
			<Portal>
				<div>I'm inside a portal</div>
			</Portal>
		);

		const { baseElement } = render(<Element />);
		expect(baseElement).toMatchSnapshot();
	});

	it(`should render a portal as custom element`, async () => {
		const Element = () => (
			<Portal type="custom-portal">
				<div>I'm inside a portal</div>
			</Portal>
		);

		const { baseElement } = render(<Element />);
		expect(baseElement).toMatchSnapshot();
	});

	it(`should append nested portals to the same ownerdocument`, async () => {
		const InnerPortal = ({ text }) => {
			return (
				<Portal>
					<div>{text}</div>
				</Portal>
			);
		};

		const Element = () => {
			return (
				<Portal>
					<div>I'm inside a portal</div>
					<InnerPortal text="inner portal"/>
				</Portal>
			);
		};

		const { baseElement } = render(<Element />);
		expect(baseElement).toMatchSnapshot();
	});
});
