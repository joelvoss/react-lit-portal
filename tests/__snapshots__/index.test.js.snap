// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portal /> should append nested portals to the same ownerdocument 1`] = `
<body>
  <div />
  <portal>
    <div>
      I'm inside a portal
    </div>
  </portal>
  <portal>
    <div>
      inner portal
    </div>
  </portal>
</body>
`;

exports[`<Portal /> should render a portal 1`] = `
<body>
  <div />
  <portal>
    <div>
      I'm inside a portal
    </div>
  </portal>
</body>
`;

exports[`<Portal /> should render a portal as custom element 1`] = `
<body>
  <div />
  <custom-portal>
    <div>
      I'm inside a portal
    </div>
  </custom-portal>
</body>
`;
